File: vinter_python/pmem.py
import collections
import dataclasses
from typing import Dict, Iterable, List, Optional, Set


def range_overlap(r1: range, r2: range) -> Optional[range]:  # assumes r1.step == r2.step == 1
    overlap = range(max(r1.start, r2.start), min(r1.stop, r2.stop))
    return overlap if len(overlap) else None


@dataclasses.dataclass(frozen=True)
class MemoryWrite:
    address_start: int
    value: bytes
    metadata: Optional[str] = None

    @property
    def address_stop(self) -> int:  # exclusive
        return self.address_start + len(self.value)

    @property
    def address_range(self) -> range:
        return range(self.address_start, self.address_stop)

    def __str__(self) -> str:
        return f'{self.address_start:#x}={self.value!r}'  # {" " + self.metadata if self.metadata else ""}

    def copy(self, *args):
        return self  # we're immutable

    __deepcopy__ = copy


class OrderedWriteLine:
    writes: List[MemoryWrite]
    flushed_index: int  # everything up until but excluding this index has been marked for flushing
    # if adding anything, update __deepcopy__

    def __init__(self):
        self.writes = []
        self.flushed_index = 0

    def flush_all(self):
        self.flushed_index = len(self.writes)

    def flushed_writes(self):
        # ~ assert(0 <= self.flushed_index <= len(self.writes))
        return self.writes[0:self.flushed_index]

    def unflushed_writes(self):
        return self.writes[self.flushed_index:]

    def clean_flushed_writes(self):
        self.writes = self.writes[self.flushed_index:]
        self.flushed_index = 0

    def __deepcopy__(self, memodict):
        # extremely improves performance
        copy = OrderedWriteLine()
        copy.writes = self.writes[:]  # MemoryWrite is immutable
        copy.flushed_index = self.flushed_index
        return copy


class X86PersistentMemory:
    memory_content: bytearray
    pending_lines: Set[int]  # just a performance improvement
    unpersisted_content: Dict[int, OrderedWriteLine]  # maps line number (== address // line_granularity) to OrderedWriteLine
    line_granularity: int

    def __init__(self, image_size: int, line_granularity: int = 64) -> None:
        assert(image_size >= 0)

        # Aligned 8-byte stores are powerfail atomic, we want to simulate at least that
        # Also, ordering of writes in cache line is guaranteed. To simulate that, choose 64.
        assert(line_granularity in (8, 64))

        self.memory_content = bytearray(image_size)
        self.pending_lines = set()
        self.unpersisted_content = collections.defaultdict(OrderedWriteLine)
        self.line_granularity = line_granularity

    def write(self, address: int, value: bytes, non_temporal: bool, metadata: str = None) -> None:
        assert(len(value) in (1, 2, 4, 8))  # test to see if we even get larger stores

        # Split into a maximum of (if possible aligned) 8-byte writes, since (aligned?!) 8 byte writes are powerfail-atomic.
        # Make sure that 8 byte boundaries are never crossed.
        address_stop = address + len(value)
        split_address_ranges = (range(max(a, address), min(a + 8, address_stop)) for a in range(address - (address % 8), address_stop if address_stop % 8 == 0 else address_stop + 8 - (address_stop % 8), 8))

        for address_range in split_address_ranges:
            line_number = address_range.start // self.line_granularity
            self.unpersisted_content[line_number].writes.append(
                    MemoryWrite(address_start=address_range.start, value=value[(address_range.start - address):(address_range.stop - address)], metadata=metadata))
            if non_temporal:  # approximation of non-temporal stores
                self.pending_lines.add(line_number)
                self.unpersisted_content[line_number].flush_all()  # note that if self.line_granularity == 64 (cache line), this is probably not quite correct

    def clwb(self, address: int, flush_writes_limit: int = None) -> None:  # flush_writes_limit is a *count*, not an index (makes a difference of 1)
        cache_line_base = (address >> 6) << 6

        for a in range(cache_line_base, cache_line_base + 64, self.line_granularity):
            # mark all writes to this address/line as "flushed"
            line_number = a // self.line_granularity
            if line := self.unpersisted_content.get(line_number):
                self.pending_lines.add(line_number)
                if flush_writes_limit:
                    line.flushed_index = flush_writes_limit
                else:
                    line.flush_all()

    def fence(self, lines: Iterable[int] = None) -> None:
        if lines is None:
            lines = self.pending_lines

        for line_number in lines:
            assert(self.unpersisted_content[line_number])
            assert(self.unpersisted_content[line_number].flushed_index > 0)
            for write in self.unpersisted_content[line_number].flushed_writes():  # go through all flushed writes in case they overlap
                self.memory_content[write.address_start:write.address_stop] = write.value
            self.unpersisted_content[line_number].clean_flushed_writes()
            assert(self.unpersisted_content[line_number].flushed_writes() == [])
            if not self.unpersisted_content[line_number].writes:
                self.unpersisted_content.pop(line_number)
            if lines is not self.pending_lines:
                self.pending_lines.remove(line_number)
        if lines is self.pending_lines:
            assert isinstance(lines, set)
            lines.clear()

    def read(self, address, size) -> bytes:
        """Reads current value from memory (independent of persistency semantics)"""

        line_number = address // self.line_granularity
        if address + size > len(self.memory_content):
            raise ValueError('address + size > len(self.memory_content)')
        content = self.memory_content[address:(address + size)]

        requested_range = range(address, address + size)
        if line := self.unpersisted_content.get(line_number):
            for write in line.writes:
                if overlap := range_overlap(requested_range, write.address_range):
                    content[(overlap.start - address):(overlap.stop - address)] = write.value[(overlap.start - write.address_start):(overlap.stop - write.address_start)]

        assert(len(content) == size)
        return content

    def persist_unpersisted(self) -> None:
        for line_number in self.unpersisted_content.keys():  # invariant: all lines are part of exactly one cache line
            self.clwb(line_number * self.line_granularity)
        self.fence()
        assert(not self.unpersisted_content)
        assert(not self.pending_lines)

    def print_unpersisted(self) -> None:
        for line_number, line in sorted(self.unpersisted_content.items()):
            print(f'unpersisted line {line_number}: {", ".join(map(str, line.writes))}')

    def print_unpersisted_verbose(self) -> None:
        for line_number, line in sorted(self.unpersisted_content.items()):
            print(f'unpersisted line {line_number} flushed from {line.flushed_index} (verbose):')
            for write in line.writes:
                print(f'{write}, metadata: {write.metadata}')

File: vinter_rust/vinter_trace2img/src/pmem.rs
use std::cmp::{max, min};
use std::collections::{HashMap, HashSet};
use std::ops::Range;

use anyhow::Result;

use crate::{MemoryImage, MemoryImageMmap};
use vinter_common::trace::Metadata;

#[derive(Debug, Clone)]
pub struct MemoryWrite {
    pub address_start: usize,
    pub value: Vec<u8>,
    pub metadata: Metadata,
}

impl MemoryWrite {
    pub fn address_end(&self) -> usize {
        self.address_start + self.value.len()
    }

    pub fn address_range(&self) -> Range<usize> {
        self.address_start..self.address_end()
    }
}

#[derive(Clone)]
pub struct OrderedWriteLine {
    writes: Vec<MemoryWrite>,
    /// everything up until but excluding this index has been marked for flushing
    flushed_index: usize,
}

impl Default for OrderedWriteLine {
    fn default() -> Self {
        Self::new()
    }
}

impl OrderedWriteLine {
    pub fn new() -> Self {
        OrderedWriteLine {
            writes: Vec::new(),
            flushed_index: 0,
        }
    }

    pub fn flush_all(&mut self) {
        self.flushed_index = self.writes.len();
    }

    pub fn all_writes(&self) -> &[MemoryWrite] {
        &self.writes
    }

    pub fn flushed_writes(&self) -> &[MemoryWrite] {
        &self.writes[0..self.flushed_index]
    }

    pub fn unflushed_writes(&self) -> &[MemoryWrite] {
        &self.writes[self.flushed_index..]
    }

    pub fn drain_flushed_writes(&mut self) -> std::vec::Drain<'_, MemoryWrite> {
        let idx = self.flushed_index;
        self.flushed_index = 0;
        self.writes.drain(0..idx)
    }

    /// Do any pending writes overlap with an access at the specified address and size?
    pub fn overlaps_access(&self, address: usize, size: usize) -> bool {
        let access_range = address..(address + size);
        self.writes
            .iter()
            .any(|w| !range_overlap(&w.address_range(), &access_range).is_empty())
    }
}

pub enum LineGranularity {
    Word,
    Cacheline,
}

impl std::convert::From<LineGranularity> for usize {
    fn from(lg: LineGranularity) -> usize {
        match lg {
            LineGranularity::Word => 8,
            LineGranularity::Cacheline => 64,
        }
    }
}

pub struct X86PersistentMemory {
    pub image: MemoryImageMmap,
    pending_lines: HashSet<usize>,
    /// maps line number (== address / line_granularity) to OrderedWriteLine
    pub unpersisted_content: HashMap<usize, OrderedWriteLine>,
    /// 8 or 64
    line_granularity: usize,
}

impl X86PersistentMemory {
    pub fn new(image: MemoryImageMmap, line_granularity: LineGranularity) -> Result<Self> {
        Ok(X86PersistentMemory {
            image,
            pending_lines: HashSet::new(),
            unpersisted_content: HashMap::new(),
            line_granularity: line_granularity.into(),
        })
    }

    /// Try to clone the memory content. Can fail due to I/O errors.
    pub fn try_clone(&self) -> Result<Self> {
        Ok(X86PersistentMemory {
            image: self.image.try_clone()?,
            pending_lines: self.pending_lines.clone(),
            unpersisted_content: self.unpersisted_content.clone(),
            line_granularity: self.line_granularity,
        })
    }

    pub fn memory_content(&self) -> &[u8] {
        &self.image
    }

    pub fn pmem_len(&self) -> usize {
        self.image.len()
    }

    pub fn line_granularity(&self) -> LineGranularity {
        match self.line_granularity {
            8 => LineGranularity::Word,
            64 => LineGranularity::Cacheline,
            other => panic!("BUG: invalid line granularity {}", other),
        }
    }

    pub fn write(&mut self, address: usize, value: &[u8], non_temporal: bool, metadata: &Metadata) {
        // test to see if we even get larger stores
        assert!(matches!(value.len(), 1 | 2 | 4 | 8));
        let address_stop = address + value.len();
        let split_address_ranges = {
            let start = address - address % 8;
            let stop = if address_stop % 8 == 0 {
                address_stop
            } else {
                address_stop + 8 - (address_stop % 8)
            };
            (start..stop)
                .step_by(8)
                .map(|a| max(a, address)..min(a + 8, address_stop))
        };

        for address_range in split_address_ranges {
            let line_number = address_range.start / self.line_granularity;
            let line = self
                .unpersisted_content
                .entry(line_number)
                .or_insert_with(OrderedWriteLine::new);
            line.writes.push(MemoryWrite {
                address_start: address_range.start,
                value: value[(address_range.start - address)..(address_range.end - address)].into(),
                metadata: metadata.clone(),
            });

            // approximation of non-temporal stores
            if non_temporal {
                self.pending_lines.insert(line_number);
                // note that for cache line granularity, this is probably not quite correct
                line.flush_all();
            }
        }
    }

    // TODO: what do we need flush_writes_limit for?
    pub fn clwb(&mut self, address: usize, flush_writes_limit: Option<usize>) {
        let cache_line_base = (address >> 6) << 6;
        for a in (cache_line_base..(cache_line_base + 64)).step_by(self.line_granularity) {
            let line_number = a / self.line_granularity;
            if let Some(line) = self.unpersisted_content.get_mut(&line_number) {
                self.pending_lines.insert(line_number);
                if let Some(limit) = flush_writes_limit {
                    line.flushed_index = limit;
                } else {
                    line.flush_all();
                }
            }
        }
    }

    pub fn fence(&mut self) {
        // A fence consumes all pending lines. Swap in a new set to avoid double borrow of self.
        let mut pending_lines = HashSet::new();
        std::mem::swap(&mut pending_lines, &mut self.pending_lines);
        for line in pending_lines {
            self.fence_line(line);
        }
    }

    pub fn fence_line(&mut self, line: usize) {
        if let Some(content) = self.unpersisted_content.get_mut(&line) {
            assert!(content.flushed_index > 0);
            for write in content.drain_flushed_writes() {
                self.image[write.address_range()].copy_from_slice(&write.value);
            }
            if content.writes.is_empty() {
                self.unpersisted_content.remove(&line);
            }
            self.pending_lines.remove(&line);
        } else {
            unreachable!();
        }
    }

    /// Reads current value from memory (independent of persistency semantics).
    /// Note: result is only correct if it fits in one line (size and alignment).
    pub fn read(&self, address: usize, size: usize) -> Vec<u8> {
        let line_number = address / self.line_granularity;
        assert!(address + size <= self.image.len());
        let range = address..(address + size);
        let mut content: Vec<u8> = self.image[range.clone()].into();

        if let Some(line) = self.unpersisted_content.get(&line_number) {
            for write in &line.writes {
                let overlap = range_overlap(&range, &write.address_range());
                if !overlap.is_empty() {
                    content[(overlap.start - address)..(overlap.end - address)].copy_from_slice(
                        &write.value[(overlap.start - write.address_start)
                            ..(overlap.end - write.address_end())],
                    );
                }
            }
        }

        content
    }

    pub fn persist_unpersisted(&mut self) {
        let lines: Vec<usize> = self.unpersisted_content.keys().copied().collect();
        for line_number in lines {
            self.clwb(line_number * self.line_granularity, None);
        }
        self.fence();
        assert!(self.unpersisted_content.is_empty());
        assert!(self.pending_lines.is_empty());
    }

    pub fn print_unpersisted(&self) {
        let mut lines: Vec<(&usize, &OrderedWriteLine)> = self.unpersisted_content.iter().collect();
        lines.sort_by_key(|(line_number, _)| *line_number);
        for (line_number, line) in lines {
            println!("unpersisted line {}: {:?}", *line_number, line.writes);
        }
    }

    /// Hash the memory contents with blake3.
    pub fn blake3(&self) -> blake3::Hash {
        blake3::hash(&self.image)
    }
}

fn range_overlap<T>(r1: &Range<T>, r2: &Range<T>) -> Range<T>
where
    T: std::cmp::Ord + Copy,
{
    Range {
        start: max(r1.start, r2.start),
        end: min(r1.end, r2.end),
    }
}
